### PR TITLE
improve accuracy of jit type signature

### PR DESCRIPTION
**Pull request by joelberkeley-secondmind**
this PR improves the accuracy of the `jit` decorator's signature

### CODE REVIEW

**Review by awav — COMMENTED** [review]
How is that different from what we have now? Why is that important?

**Comment by joelberkeley-pio** [comment]
the type signature `def jit(apply: bool = True, **optimize_kwargs) -> Callable[[Callable], Callable]:` loses information about the function it decorates, because `Callable` [is equivalent to](https://docs.python.org/3/library/typing.html#typing.Callable) `Callable[..., Any]`. For example,
```
@jit()
def foo(x: int) -> str:
    ...
```
would be untyped

File: trieste/utils/misc.py [hunk 1]
```diff
@@ -11,15 +11,18 @@
...
 import tensorflow as tf
 
 from ..type import TensorType
 
+C = TypeVar("C", bound=Callable)
+""" A type variable bound to `typing.Callable`. """
 
-def jit(apply: bool = True, **optimize_kwargs) -> Callable[[Callable], Callable]:
+
+def jit(apply: bool = True, **optimize_kwargs) -> Callable[[C], C]:
```
Comment (joelberkeley-secondmind) [inline trieste/utils/misc.py:25]: i'm leaving sorting out the argument types for another time

**Review by ChrisMorter — APPROVED** [review]
Looks good to me

**Pull request merged**